src/vsr/grid.zig

```diff
@@ -765,76 +892,57 @@ pub fn GridType(comptime Storage: type) type {
             // Remove the "root" read so that the address is no longer actively reading / locked.
             grid.read_queue.remove(read);
 
-            // A valid block filled by storage means the reads for the address can be resolved.
-            if (read_block_valid(cache_block.*, .{
+            const result = read_block_validate(block.*, .{
                 .address = read.address,
                 .checksum = read.checksum,
-                .block_type = read.block_type,
-            })) {
-                grid.read_block_resolve(read, cache_block.*);
+            });
+
+            if (result == .valid) {
+                grid.read_block_resolve(read, .{ .valid = block.* });
                 return;
             }
 
+            const header = mem.bytesAsValue(vsr.Header, block.*[0..@sizeOf(vsr.Header)]);
+            log.err(
+                "{s}: expected address={} checksum={}, found address={} checksum={}",
+                .{
+                    @tagName(result),
+                    read.address,
+                    read.checksum,
+                    header.op,
+                    header.checksum,
+                },
+            );
+
             // Don't cache a corrupt or incorrect block.
             grid.cache.remove(read.address);
-            grid.cache_coherent.unset(cache_index);
 
-            // On the result of an invalid block, move the "root" read (and all others it
-            // resolves) to recovery queue. Future reads on the same address will see the "root"
-            // read in the recovery queue and enqueue to it.
-            grid.read_faulty_queue.push(read);
-            if (grid.on_read_fault) |on_read_fault| {
-                on_read_fault(grid, read);
-            } else {
-                @panic("Grid.on_read_fault not set");
-            }
+            grid.read_block_resolve(read, result);
         }
 
-        fn read_block_valid(block: BlockPtrConst, expect: struct {
+        fn read_block_validate(block: BlockPtrConst, expect: struct {
             address: u64,
             checksum: u128,
-            block_type: ?BlockType,
-        }) bool {
-            const address = expect.address;
-            const checksum = expect.checksum;
-            const block_type = expect.block_type;
-
+        }) ReadBlockResult {
             const header = mem.bytesAsValue(vsr.Header, block[0..@sizeOf(vsr.Header)]);
 
-            if (!header.valid_checksum()) {
-                log.err("invalid checksum at address {} (expected={})", .{ address, checksum });
-                return false;
-            }
+            if (!header.valid_checksum()) return .invalid_checksum;
+            if (header.command != .block) return .unexpected_command;
```

Comment:
Sneaky bug fix here!

We need to verify the command before we validate the body checksum.

If this is a misdirected `command=prepare` or `command=reply`, its header is valid (`valid_checksum()`) -- but checking the body checksum could panic! Blocks (and `block`) have a 64KiB capacity, but prepares/replies may be as large as 1MiB.